app/models/manageiq/providers/azure/cloud_manager/provision/cloning.rb

```diff
@@ -0,0 +1,143 @@
+
+      return true if status == "Succeeded"
+      return false, status
+    end
+  end
+
+  def find_destination_in_vmdb(vm_uid)
+    ems_ref = vm_uid.join("\\")
+    ManageIQ::Providers::Azure::CloudManager::Vm.find_by(:ems_ref => ems_ref.downcase)
+  end
+
+  def gather_storage_account_properties
+    sas = nil
+    source.with_provider_connection do |azure|
+      sas = Azure::Armrest::StorageAccountService.new(azure)
+    end
+    return if sas.nil?
+
+    begin
+      key             = sas.list_account_keys(storage_account_name, storage_account_resource_group).fetch('key1')
+      storage_account = sas.get(storage_account_name, storage_account_resource_group)
+      blob            = storage_account.blobs("system", key).first
+      blob_properties = storage_account.blob_properties(blob.container, blob.name, key)
+      endpoint        = storage_account.properties.primary_endpoints.blob
+      source_uri      = File.join(endpoint, blob.container, blob.name)
+      target_uri      = File.join(endpoint, "manageiq", dest_name + "_" + SecureRandom.uuid + ".vhd")
+    rescue Azure::Armrest::ResourceNotFoundException => err
+      _log.error("Error Class=#{err.class.name}, Message=#{err.message}")
+    end
+
+    return target_uri, source_uri, blob_properties.x_ms_meta_microsoftazurecompute_ostype
+  end
+
+  def root_password
+    MiqPassword.decrypt(options[:root_password]) if options[:root_password]
+  end
+
+  def custom_data
+    Base64.encode64(userdata_payload.encode("UTF-8")).delete("\n")
+  end
+
+  def prepare_for_clone_task
+    nic_id = create_nic
+    target_uri, source_uri, os = gather_storage_account_properties
+
+    cloud_options = {
+      :name       => dest_name,
+      :location   => source.location,
+      :properties => {
+        :hardwareProfile => {
+          :vmSize => instance_type.name
+        },
+        :osProfile => {
+            :adminUserName => options[:root_username],
+            :adminPassword => root_password,
+            :computerName  => dest_name
+        },
+        :storageProfile => {
+          :osDisk => {
+            :createOption => 'FromImage',
+            :caching      => 'ReadWrite',
+            :name         => dest_name + SecureRandom.uuid + '.vhd',
+            :osType       => os,
+            :image        => {:uri => source_uri},
+            :vhd          => {:uri => target_uri},
+          }
+        },
+        :networkProfile => {
+          :networkInterfaces => [{:id => nic_id}],
+        }
+      }
+    }
+    cloud_options[:properties][:osProfile][:customData] = custom_data unless userdata_payload.nil?
+    cloud_options
+  end
+
+  def log_clone_options(clone_options)
+    dumpObj(clone_options, "#{_log.prefix} Clone Options: ", $log, :info)
+    dumpObj(options, "#{_log.prefix} Prov Options:  ", $log, :info, :protected =>
+    {:path => workflow_class.encrypted_options_field_regs})
+  end
+
+  def region
+    source.location
+  end
+
+  def resource_group
```

Comment:
I would expect this method to be in a [options_helper.rb](https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/cloud_manager/provision/options_helper.rb#L1) file for Azure.  Could also move the `root_password` method in there as well.